packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts

```diff
@@ -151,6 +153,56 @@ export async function createGraphqlEngineBundle(
 
   return new Promise((resolve, reject) => {
     compiler.run((err, stats) => {
+      function getResourcePath(
+        webpackModule?: Module | NormalModule | ConcatenatedModule | null
+      ): string | undefined {
+        if (!(webpackModule instanceof ConcatenatedModule)) {
+          return (webpackModule as NormalModule).resource
+        }
+
+        // ConcatenatedModule is a collection of modules so we have to go deeper to actually get a path,
+        // at this point we won't know which one so we just grab first module here
+        const [firstSubModule] = webpackModule.modules
+        return getResourcePath(firstSubModule)
+      }
+
+      let tsNodeUsed = false
+      function iterateModules(
+        webpackModules: Set<Module>,
+        compilation: Compilation
+      ): void {
+        for (const webpackModule of webpackModules) {
+          if (tsNodeUsed) {
+            // once we find ts-node module anywhere we can stop iterating
+            break
+          }
+
+          if (webpackModule instanceof ConcatenatedModule) {
+            iterateModules(
+              (webpackModule as ConcatenatedModule).modules,
+              compilation
+            )
+          } else {
+            const resourcePath = getResourcePath(webpackModule)
+            if (resourcePath?.includes(`ts-node`)) {
+              const importedBy = getResourcePath(
+                compilation.moduleGraph.getIssuer(webpackModule)
+              )
+              console.warn(
+                `"ts-node" usage detected${
+                  importedBy ? ` (imported by ${importedBy})` : ``
+                }`
+              )
```

Comment:
This also might be panic here. Current versions already don't work when `ts-node` is used due to `typescript` problem when validating, so panicking here won't change that, but we could have concrete error message and mitigation (removal of `ts-node` as it's not needed anymore with builtin TS support - this would be benificial not only for engine usage, but just in general - all the `ts-node/register` can increase memory problems and that's one of the reason built in support in Gatsby doesn't use require hooks and instead precompile those files)

Reply:
lets panic with a structured log pointing to the docs

Reply:
After additional thought - we are already failing today when someone is using `ts-node` somewhere in the code that we bundle for engines, so same builds that fail today will continue to fail, just with nicer and actually actionable message. 

In general `ts-node` is problematic and not needed anymore (not just for engines, you will find mentions of increased memory usage, which is particularly painful for us even in regular build), so I did change my mind on this after zooming out of this concrete problem of handling ts-node in engines